pkg/supervisor/supervisor.go

```diff
@@ -0,0 +1,247 @@
+// invoking supervisor.Shutdown(). This can be done from any goroutine
+// including workers.
+//
+// The graceful shutdown sequence shuts down workers in an order that
+// respects worker dependencies.
+func (s *Supervisor) Run() []error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	for len(s.workers) > 0 {
+		s.reconcile()
+		s.changed.Wait()
+	}
+	return s.errors
+}
+
+// Triggers a graceful shutdown sequence. This can be invoked from any
+// goroutine.
+func (s *Supervisor) Shutdown() {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	s.shuttingDown = true
+	s.changed.Broadcast()
+}
+
+func (s *Supervisor) dependents(worker *Worker) (result []*Worker) {
+	for _, n := range s.names {
+		w := s.workers[n]
+		for _, r := range w.Requires {
+			if r == worker.Name {
+				result = append(result, w)
+				break
+			}
+		}
+	}
+	return
+}
+
+// make sure anything that would like to be running is actually
+// running
+func (s *Supervisor) reconcile() {
+	if s.shuttingDown {
+		s.reconcileShutdown()
+	} else {
+		s.reconcileNormal()
+	}
+}
+
+// shutdown anything that is safe to shutdown and not already shutdown
+func (s *Supervisor) reconcileShutdown() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process != nil && !w.process.shutdownClosed {
+			for _, d := range s.dependents(w) {
+				if s.workers[d.Name].process != nil {
+					log.Printf("cannot shutdown %s, %s still running", n, d.Name)
+					continue OUTER
+				}
+			}
+			close(w.process.shutdown)
+			w.process.shutdownClosed = true
+		}
+	}
+}
+
+// launch anything that is safe to launch and not already launched
+func (s *Supervisor) reconcileNormal() {
+OUTER:
+	for _, n := range s.names {
+		w := s.workers[n]
+		if w.process == nil {
+			for _, r := range w.Requires {
+				process := s.workers[r].process
+				if process == nil || !process.ready {
+					log.Printf("cannot start %s, %s not ready", n, r)
+					continue OUTER
+				}
+			}
+			s.launch(w)
+		}
+	}
+}
+
+func (s *Supervisor) launch(worker *Worker) {
+	context, cancel := context.WithCancel(s.context)
+	process := &Process{
+		Supervisor: s,
+		Worker:     worker,
+		Context:    context,
+		cancel:     cancel,
+		shutdown:   make(chan struct{}),
+	}
+	worker.process = process
+	go func() {
+		var err error
+		func() {
+			defer func() {
+				if r := recover(); r != nil {
+					err = errors.Errorf("PANIC: %v", r)
+				}
+			}()
+			err = worker.Work(process)
+		}()
+		s.mutex.Lock()
+		defer s.mutex.Unlock()
+		worker.process = nil
+		if err != nil {
+			process.Log(err)
+			if worker.Retry {
+				if s.shuttingDown {
+					s.remove(worker)
+				} else {
+					process.Log("retrying...")
+				}
+			} else {
+				s.remove(worker)
+				s.errors = append(s.errors, err)
+				s.shuttingDown = true
+			}
+		} else {
+			s.remove(worker)
+		}
+		s.changed.Broadcast()
+	}()
+}
+
+type Process struct {
+	Supervisor *Supervisor
+	Worker     *Worker
+	// Used for hard cancel.
+	Context context.Context
```

Comment:
I'd consider making these private, and exposing them in a read-only way via getter methods, like you did for `shutdown`.

I'm not sure what the use-case is for exposing the `Worker` to the `Work` func.

Reply:
I've done as you suggested. The use case for exposing the Worker from the Process object was to access the name when logging. Strictly speaking that doesn't need to be public as-is, but with a pluggable logger I think it would actually need to be public.